src/com/google/javascript/jscomp/Es6ConvertSuper.java

```diff
@@ -148,18 +184,56 @@ private void visitSuper(Node node, Node parent) {
 
     String methodName;
     Node callName = enclosingCall.removeFirstChild();
+    Node superThis = IR.name(SUPER_THIS);
+    superThis.setOriginalName("this");
     if (callName.isSuper()) {
       methodName = enclosingMemberDef.getString();
+      if (methodName.equals("constructor")) {
+        Node ctorFunction = enclosingMemberDef.getFirstChild();
+        Node body = NodeUtil.getFunctionBody(ctorFunction);
+        NodeTraversal.traverseEs6(compiler, body, new UpdateThisReferences());
+        body.addChildToBack(IR.returnNode(superThis.cloneNode()));
+      }
     } else {
       methodName = callName.getLastChild().getString();
     }
+
     Node baseCall = baseCall(
         superName.getQualifiedName(), methodName, enclosingCall.removeChildren());
+    Node nodeToReplace = enclosingCall;
+    if (methodName.equals("constructor")) {
+      baseCall = IR.var(
+          superThis.cloneNode(),
+          IR.or(baseCall, IR.thisNode()));
+      JSDocInfoBuilder info = new JSDocInfoBuilder(false);
+      info.recordType(new JSTypeExpression(
+          new Node(Token.BANG, IR.string(className)), clazz.getSourceFileName()));
+      info.recordConstancy();
+      baseCall.setJSDocInfo(info.build());
+      nodeToReplace = enclosingCall.getParent();
+      Preconditions.checkState(nodeToReplace.isExprResult());
+    }
     baseCall.useSourceInfoIfMissingFromForTree(enclosingCall);
-    enclosingCall.getParent().replaceChild(enclosingCall, baseCall);
+    nodeToReplace.getParent().replaceChild(nodeToReplace, baseCall);
     compiler.reportCodeChange();
   }
 
+  private static class UpdateThisReferences implements NodeTraversal.Callback {
+    @Override
+    public void visit(NodeTraversal t, Node n, Node parent) {
+      if (n.isThis()) {
```

Comment:
So then this would need to be `n.isThis() || "this".equals(n.getOriginalName())`. Are there any other `isThis` calls that need to change?